### ISSUE TITLE
{case_clause,{ok,error}} on Pub and -normalize_subscribe_topics/1-fun-0- on sub

### ISSUE THREAD

**Issue by xbugster** [vernemq/vernemq#958]
### Environment

- VerneMQ Version: 1.5.0-127-ge130553
- OS: ubuntu linux 16.04
- Erlang/OTP version (if building from source): erts-8.3.5.5@OTP 19

### Expected behavior

First, worth reminding, we utilize Lua plugin support.
We actually override the topic from /request to /response (as we mimicking http over mqtt).
On publish, we encounter ```[error] CRASH REPORT Process <0.527.0> with 0 neighbours crashed with reason: no case clause matching {ok,error} in vmq_mqtt_fsm:auth_on_publish/4 line 669
[error] Ranch listener {{172,17,0,2},1883} terminated with reason: {case_clause,{ok,error}}```
Which seems to come from core vernemq code, we also receive error on subscribe, we perform nothing except authenticating client on our side. we return parameters unchanged, but, we receive ```[error] CRASH REPORT Process <0.519.0> with 0 neighbours crashed with reason: no function clause matching vmq_diversity_utils:'-normalize_subscribe_topics/1-fun-0-'({client_id,<<"mosqsub|22522-bountyhun">>}) line 38
[error] Ranch listener {{172,17,0,2},1883} terminated with reason: function_clause```

--

We will provide more information. The topic that message should be published to (printed just before the pub table returned in auth_on_publish():
`[info] PUB AUTH BEFORE RETURN: v1/ingest/response/device/id`

And the topics value for subscribe is (also printed before return on auth_on_subscribe with json.encode(value)):
`[info] SUB AUTH BEFORE RETURN: [["v1/ingest/response/device/id",0.0]]`

### Actual behaviour

- We expect everything to work as described in documentation. If we change topic for publish within auth_on_publish() hook, we expect payload being published to a new topic.

- During subscription, we expect client to successfully subscribe (no data manipulations done there).

**Comment by xbugster** [comment]
Version 1.6.1:
Subscribe error:
```[error] CRASH REPORT Process <0.469.0> with 0 neighbours crashed with reason: no function clause matching vmq_diversity_utils:'-normalize_subscribe_topics/1-fun-0-'({client_id,<<"mosqsub|25322-bountyhun">>}) line 38
[error] Ranch listener {{172,17,0,2},1883} terminated with reason: no function clause matching vmq_diversity_utils:'-normalize_subscribe_topics/1-fun-0-'({client_id,<<"mosqsub|25322-bountyhun">>}) line 38```

Publish error:
```[error] CRASH REPORT Process <0.489.0> with 0 neighbours crashed with reason: no case clause matching {ok,error} in vmq_mqtt_fsm:auth_on_publish/4 line 669
[error] Ranch listener {{172,17,0,2},1883} terminated with reason: no case clause matching {ok,error} in vmq_mqtt_fsm:auth_on_publish/4 line 669```

**Comment by xbugster** [comment]
Full list of value on publish (just before return on auth):
`{"client_id":"xbugsterasd111","mountpoint":"","payload":"{\"id\":\"any_request_id123312\",\"status\":404.0}","qos":0.0,"retain":false,"topic":"v1/ingest/response/device/id","username":"otoken=ftvIO9cDUhSkYCeomKdxkA7ggv6lmIQE"}`

Full list of values for subscribe:
`{"client_id":"mosqsub|26171-bountyhun","mountpoint":"","topics":[["v1/ingest/response/device/id",0.0]],"username":"otoken=ftvIO9cDUhSkYCeomKdxkA7ggv6lmIQE"}`

**Comment by dergraf** [comment]
It looks like the modifiers you've provided as part of your hook return value aren't correct.

But, it's a bug that vmq_diversity returns {ok, error} in this case (modifier validation failed), we should return an {error, Reason} instead.

**Comment by xbugster** [comment]
Thank You. It seems, we changed 2 values during publish (the payload and topic), now it tries to access auth_on_publish/4 rather than /6. o.O method doesn't exist ? perhaps...
seems to be the same with subscribe, as we played with topic, but not changed it.

**Comment by larshesel** [comment]
Can you provide the lua script snippet with the return values that cause this?

**Comment by xbugster** [comment]
```lua
function my_auth_on_publish(pub)
    -- following call returns modified message content as json string
    pub.payload = process_message(pub.topic, pub.payload) 
    pub.topic = string.gsub(pub.topic, '/request/', '/response/')
    return pub
end
```

and subscribe is:
```lua
function my_auth_on_subscribe(sub)
    -- following is internal ACL which parses topic, but not changes it. (splits topic by "/")
    -- we do not support multiple topics subscribe
    if(false == on_auth_init(msg_directions['subscribe'], sub.username, sub.topics[1][1])) then
        return false
    end

    return sub
end
```

**Comment by larshesel** [comment]
OK, that won't work as only specific things can be modified. In for example the subscribe case you're also passing back the username which is not a legal modifier.

**Comment by xbugster** [comment]
This means i have to structure new return object which drops off unnecessary items ?

**Comment by larshesel** [comment]
yes, here's an example from our test-suites: https://github.com/vernemq/vernemq/blob/master/apps/vmq_diversity/test/plugin_test.lua#L70

**Comment by xbugster** [comment]
Cool! Thank you. just wanted to ask you to describe the structures expected to be returned by auth_on_* 
I will try it soon and update the ticket on the outcome.
